### PR TITLE
Return inserted/replaced paths

### LINKED ISSUE #4363: `if for` doesn't capture `this` for function inside loop without curly braces. (T7449)
> Issue originally made by @alexkg
### Bug information
- **Babel version:** 6.9.1
- **Node version:** 5.4.1
- **npm version:** 3.3.12
### Input code

```
const kv = {a:1}

function WithoutCurlyBraces() {
    if (true)
      for (let k in kv) {
          function foo() { return this }
          function bar() { return foo.call(this) }
          console.log(bar.call(this)) // => undefined
      }
}

WithoutCurlyBraces.call(666)

function WithCurlyBraces() {
    if (true) {
      for (let k in kv) {
          function foo() { return this }
          function bar() { return foo.call(this) }
          console.log(bar.call(this)) // => 777
      }
    }
}

WithCurlyBraces.call(777)

```

### PR DESCRIPTION
**Pull request by jridgewell**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | Yes
| Minor: New Feature?      | Yes?
| Deprecations?            | 
| Spec Compliancy?         | 
| Tests Added/Pass?        | 
| Fixed Tickets            | Fixes #4363
| License                  | MIT
| Doc PR                   |
| Dependency Changes       | 

This gives `Path`’s replacement and insertion methods a consistent
return value: the inserted/replaced paths.

Before, they could return `undefined`, a `node`, or a the current path
inside an array. It was kinda pointless.  But now they always return an
array of paths, which is useful for solving
https://github.com/babel/babel/pull/4935#discussion_r96151368.

### CODE REVIEW

**Comment by danez** [comment]
Could you rebase your change? Then I can test it in the repl 😀

File: packages/babel-traverse/src/path/modification.js [hunk 1]
```diff
@@ -24,20 +24,18 @@ export function insertBefore(nodes) {
...
   ) {
     if (this.node) nodes.push(this.node);
-    this.replaceExpressionWithStatements(nodes);
+    return this.replaceExpressionWithStatements(nodes);
   } else if (Array.isArray(this.container)) {
     return this._containerInsertBefore(nodes);
   } else if (this.isStatementOrBlock()) {
     if (this.node) nodes.push(this.node);
-    this._replaceWith(t.blockStatement(nodes));
+    return this._replaceWith(t.blockStatement(nodes));
```
Comment (danez) [inline packages/babel-traverse/src/path/modification.js]: I think in this case were we implicitly create a `BlockStatement` we should not return the block statement but instead only the inserted nodes. I was thinking about what the usecase could be  that someone might want to know if this was created, but I think it does not have any value to return the BlockStatement itself.

Assuming this two examples

```js
if (true) console.log('');

if (true) {
  console.log('');
}
```

And you always want to insert a Statement before or after `console.log()` I think it should be consistent and always return the path to the nodes that were inserted.
Right now `inspertBefore/After()` would in this two cases still return different results.

**Comment by danez** [comment]
@jridgewell I hope it is okay, I added the fix I did in #6160 to this PR as it falls into the same category. Now `insertBefore` and `insertAfter` will return the inserted paths also when a `BlockStatement` is created. Also added the testcase for #4363 which works now.

**Comment by jridgewell** [comment]
Thanks for the tests. Did a cleanup, should be good for final review.

**Comment by hzoo** [comment]
What's the changelog entry/thing we write for the upgrade post for plugin authors?

**Pull request merged**